### ISSUE TITLE
False Positive (FP) on 'background' does not affect IoU calculation.

### ISSUE THREAD

**Issue by csj3684** [open-mmlab/mmsegmentation#3340]
I really appreciate your works. :)

I have some questions about code 'def intersect_and_union' in 'mmsegmentation/mmseg/evaluation/metrics/iou_metric.py'

    def intersect_and_union(pred_label: torch.tensor, label: torch.tensor, num_classes: int, ignore_index: int):
        """Calculate Intersection and Union.

        Args:
            pred_label (torch.tensor): Prediction segmentation map
                or predict result filename. The shape is (H, W).
            label (torch.tensor): Ground truth segmentation map
                or label filename. The shape is (H, W).
            num_classes (int): Number of categories.
            ignore_index (int): Index that will be ignored in evaluation.

        Returns:
            torch.Tensor: The intersection of prediction and ground truth
                histogram on all classes.
            torch.Tensor: The union of prediction and ground truth histogram on
                all classes.
            torch.Tensor: The prediction histogram on all classes.
            torch.Tensor: The ground truth histogram on all classes.
        """

        mask = (label != ignore_index)            <-----------  mask by 'GT'
        pred_label = pred_label[mask]            <-----------  FP on background in model output is ignored.
        label = label[mask]

        intersect = pred_label[pred_label == label]
        area_intersect = torch.histc(intersect.float(), bins=(num_classes), min=0, max=num_classes - 1).cpu()
        area_pred_label = torch.histc(pred_label.float(), bins=(num_classes), min=0, max=num_classes - 1).cpu()
        area_label = torch.histc(label.float(), bins=(num_classes), min=0, max=num_classes - 1).cpu()
        area_union = area_pred_label + area_label - area_intersect

        return area_intersect, area_union, area_pred_label, area_label

here, 'mask' variable in first line is pixels that have specific target class (not background or etc) in 'GT' mask

why masking pred_label (model output) by that mask ?  then when calculating IoU, False Positive (FP) on background in 'output' mask do not affect IoU calculation. so IoU will be higher than real IoU. 

for instance, When GT target class area is 100 and model output of target class is 120 with completely covering 'GT' area, IoU results will be 100 (%).

Thank you again for your fantastic works. XD

**Comment by csj3684** [comment]
I understood the meaning of 'ignore_index'.

**Comment by qmh045** [comment]
hi there! I meet the same condition in this IoU as yours. "Union" as "Pred∪Gt", while IoU in mmseg code here is seems like to be "P∩G / G", not the "P∩G / P∪G" we are wanting?

**Comment by qmh045** [comment]
> hi there! I meet the same condition in this IoU as yours. "Union" as "Pred∪Gt", while IoU in mmseg code here is seems like to be "P∩G / G", not the "P∩G / P∪G" we are wanting?

I realized something XD. Considering "ignore_index", if a position is 0, which is an area that should not be predicted to be a non-zero number, so it CAN'T be set to ignore,wright?  An area should only be set to ignore is that we DON'T CARE about it at all. Foolish as me

**Comment by Jason-Fu-git** [comment]
I meet the same problem too, and I think @qmh045 is right. This `ignore_index` can be misleading. You cannot mark the background class as ignored, because you care about false positive on it. The contour annotation of objects in PASCAL VOC 2012, for example, is what should be marked as ignored.
